[LBSE] Add RenderSVGResourceClipper

rwlbuis opened the pull request:
#### 57393b599e9588c477ffce15a7a729ddbeceea61
<pre>
[LBSE] Add RenderSVGResourceClipper
<a href="https://bugs.webkit.org/show_bug.cgi?id=262191">https://bugs.webkit.org/show_bug.cgi?id=262191</a>

Reviewed by Nikolas Zimmermann.

Implement the first SVG resource in LBSE: clipping (RenderSVGResourceClipper).
Unlike the legacy engine, LBSE does not make use of temporary ImageBuffers for clipping,
but relies on Porter-Duff DestinationIn/SourceOver compositing operations, to achieve
the same effect, but faster. This aligns SVG clipping with HTML/CSS clipping.

This is only the beginning: dynamic invalidations are not working, as we&apos;re introducing
them incrementally in a unified, clean way.

Covered by existing tests (dozens of progressions, see TestExpectations).

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-32-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/coords-viewattr-03-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-05-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-06-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-kern-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-intro-01-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-02-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-04-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-05-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/painting-marker-02-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/shapes-intro-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-align-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-05-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-06-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textEffect2-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textProperties-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/carto.net/textbox-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/embedding-external-svgs-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/path-moveto-only-rendering-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/subpaths-moveto-only-rendering-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/text-clip-expected.png: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/text-clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-symbol-inside-pattern-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/filters/filter-clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-align-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-text-05-t-expected.txt:
* LayoutTests/svg/clip-path/clip-opacity.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::removeSVGResource):
(WebCore::TreeScope::lookupSVGResourceById const):
(WebCore::TreeScope::lookupLegacySVGResoureById const):
(WebCore::TreeScope::svgResourceById const): Deleted.
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::hitTestClipPath const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::updateAncestorDependentState):
(WebCore::RenderLayer::computeHasVisibleContent const):
(WebCore::RenderLayer::beginTransparencyLayers):
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::setupClipPath):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::enclosingSVGRootLayer const):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::isPaintingSVGResourceLayer const):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasFilter const):
(WebCore::RenderLayer::hasNonOpacityTransparency const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGForeignObject const):
(WebCore::RenderObject::isSVGResourceContainer const):
(WebCore::RenderObject::isLegacySVGResourceClipper const):
(WebCore::RenderObject::isSVGResourceClipper const):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::computeClipPath const):
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGResource.cpp:
(WebCore::requestPaintingResource):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp: Added.
(WebCore::RenderSVGResourceClipper::RenderSVGResourceClipper):
(WebCore::RenderSVGResourceClipper::shouldApplyPathClipping const):
(WebCore::currentClippingMode):
(WebCore::sharedClipAllPath):
(WebCore::RenderSVGResourceClipper::applyPathClipping):
(WebCore::RenderSVGResourceClipper::applyMaskClipping):
(WebCore::RenderSVGResourceClipper::hitTestClipContent):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceClipperInlines.h: Added.
(WebCore::RenderSVGResourceClipper::clipPathElement const):
(WebCore::RenderSVGResourceClipper::clipPathUnits const):
* Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp:
(WebCore::RenderSVGResourceSolidColor::applyResource):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::paintSVGClippingMask):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::isRenderingMaskImage):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::drawContentIntoMaskImage):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.h:
(isType):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
(WebCore::getRenderSVGResourceContainerById):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::createElementRenderer):
(WebCore::SVGClipPathElement::shouldApplyPathClipping const):
* Source/WebCore/svg/SVGClipPathElement.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):

Canonical link: <a href="https://commits.webkit.org/269635@main">https://commits.webkit.org/269635@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43b2050d89fff1c199e9902a0224272c7bf3379c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23124 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1207 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24243 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25043 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21385 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23391 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2594 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23662 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22242 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23364 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/746 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25892 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20940 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27107 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21015 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21204 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24976 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/628 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18407 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/558 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20717 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5516 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/832 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/dom/TreeScope.h:
```diff
@@ -131,9 +132,11 @@ class TreeScope {
     void removeSVGResource(const AtomString& id);
-    LegacyRenderSVGResourceContainer* svgResourceById(const AtomString& id) const;
+    RenderSVGResourceContainer* svgResourceById(const AtomString& id) const;
+    LegacyRenderSVGResourceContainer* legacysvgResourceById(const AtomString& id) const;
```
How about:
lookupSVGResourceById and lookupLegacySVGResoureById?

rwlbuis replied:
Sounds better, done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1549,6 +1550,10 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
             break;
         if (!is<SVGClipPathElement>(*element))
             break;
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
```
I'd move this section into a new helper lambda to avoid cluttering the main logic with ifdefs and such.

```
auto hitTestClipContent = [&] -> bool {
#if ENABLE(...)
    if (auto* clipper = ...)
        return clipper->...
#else
    return somethingElse;
#endif
};
```

rwlbuis replied:
Done, but I think we need the fallback into legacy handling in case LBSE ifdef is true but the feature is not runtime enabled.

nikolaszimmermann replied:
Hmm, I don't think this adds readability unless you move the lambda out of the for loop.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -529,6 +530,8 @@ void RenderLayer::insertOnlyThisLayer(LayerChangeTiming timing)
     // Clear out all the clip rects.
     clearClipRectsIncludingDescendants();
+
+    m_enclosingSVGResourceContainer = lineageOfType<RenderSVGResourceContainer>(renderer()).first();
```
I added this in LBSE downstream for performance reasons, to avoid repeated ancestor walks.
However I was never sure, if this is the proper point to cache this, or if this is safe this way. We should be careful and revisit this.

nikolaszimmermann replied:
Revisited and found out that this place is incorrect -- moved the caching of m_enclosingSVGResourceContainer / m_enclosingSVGHiddenContainer to RenderLayer::updateAncestorDependantState() -- we're querying the ancestor chain after all, so this is the natural place.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
     return false;
 }
 
+void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const LayoutRect& paintDirtyRect, const AffineTransform& layerContentTransform)
```
Not sure if the naming is clear. A "SVG resource layer" is a RenderLayer, that is never painted directly but only indirectly via a "resource" (e.g. RenderSVGResourceClipper that wants to paint the _content_ (aka. the children) of the RenderLayer associated with e.g. the \<clipPath\>.

nikolaszimmermann replied:
Cannot think of a better name right now, let's resolve the item for now and postpone this for another review round.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
 
+void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const LayoutRect& paintDirtyRect, const AffineTransform& layerContentTransform)
+{
+    // Requires the caller to guard the context with a GraphicsContextStateSaver.
```
We apparently have better mechanism to enforce this nowadays (GraphicsContextStateSaver& as input parameter).

rwlbuis replied:
Right, implemented, but I had to add the save, I think it was wrongly omitted before.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
+void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const LayoutRect& paintDirtyRect, const AffineTransform& layerContentTransform)
+{
+    // Requires the caller to guard the context with a GraphicsContextStateSaver.
+    bool wasPaintingSVGResourceLayer = m_isPaintingSVGResourceLayer;
```
SetForScope<> avoids this dance.

rwlbuis replied:
Actually it is a bitfield :]

nikolaszimmermann replied:
Ok, so it needs an inlined getter/setter to work with SetForScope?

rwlbuis replied:
I don't think that will work either.

nikolaszimmermann replied:
Correct, there is no variant of SetForScope<> that takes a setter/getter -- mixing up projects here.
Let's leave it for now.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
+    m_isPaintingSVGResourceLayer = true;
+    context.concatCTM(layerContentTransform);
+
+    bool hasFinitePaintDirtyRect = paintDirtyRect != LayoutRect::infiniteRect();
```
Usually HTML/CSS code expects finite paint rects -- we use infiniteRect() as special case in LBSE to indicate a paint request that is not connected with the on-screen painting (e.g. to paint a whole subtree during clipping with a certain compositing operation). It deserves a comment :-)

rwlbuis replied:
I think we maybe even do not need the paintDirtyRect parameter, WDYT?

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
+
+    bool hasFinitePaintDirtyRect = paintDirtyRect != LayoutRect::infiniteRect();
+
+    LayoutRect localPaintDirtyRect = paintDirtyRect;
```
s/LayoutRect/auto/ ?

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2909,6 +2912,34 @@ static inline bool shouldSuppressPaintingLayer(RenderLayer* layer)
+
+    LayoutRect localPaintDirtyRect = paintDirtyRect;
+    if (hasFinitePaintDirtyRect)
+        localPaintDirtyRect = LayoutRect(encloseRectToDevicePixels(layerContentTransform.inverse().value_or(AffineTransform()).mapRect(paintDirtyRect), renderer().document().deviceScaleFactor()));
```
Maybe worth to write this in a more readable way?

rwlbuis replied:
Yeah, too long :) I made an attempt.

nikolaszimmermann replied:
Looks good.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3093,7 +3124,7 @@ std::pair<Path, WindRule> RenderLayer::computeClipPath(const LayoutSize& offsetF
 }
 
-void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSaver& stateSaver, RegionContextStateSaver& regionContextStateSaver, const LayerPaintingInfo& paintingInfo, OptionSet<PaintLayerFlag> paintFlags, const LayoutSize& offsetFromRoot)
+void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSaver& stateSaver, RegionContextStateSaver& regionContextStateSaver, const LayerPaintingInfo& paintingInfo, OptionSet<PaintLayerFlag> paintFlags, const LayoutSize& offsetFromRoot, bool& paintSVGClippingMask)
```
I dislike passing in a bool reference. We can maybe compute the information prior to calling setupClipPath (e.g. setupSVGClipPath) and extend 'PaintLayerFlag' with 'PaintingSVGClippingMask'? This would save another boolean-flag parameter.

nikolaszimmermann replied:
This would be handy to implement.

rwlbuis replied:
Done, but it does mean paintFlags param gains a reference.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3103,6 +3134,15 @@ void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSa
     if (is<LegacyRenderSVGRoot>(renderer()))
         return;
 
+    // Applying clip-path on <clipPath> enforces us to use mask based clipping, so return false here to disable path based clipping.
```
That logic could love outside this method...

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3124,6 +3164,36 @@ void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSa
+            FloatSize coordinateSystemOriginTranslation;
+            if (renderer().isSVGLayerAwareRenderer()) {
+                ASSERT(paintingInfo.subpixelOffset.isZero());
+                auto boundingBoxTopLeftCorner = flooredLayoutPoint(renderer().objectBoundingBox().minXMinYCorner());
```
Be careful, I guess we need to use the 'objectBoundingBoxIgnoringTransformations' here.
The better idiom is to use "nominalSVGLayoutLoction" (grep for that in rendering/svg).

rwlbuis replied:
Sure, fixed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3410,11 +3479,11 @@ void RenderLayer::paintLayerContents(GraphicsContext& context, const LayerPainti
         }
 
-        if (!(paintFlags & PaintLayerFlag::PaintingCompositingMaskPhase) && (paintFlags & PaintLayerFlag::PaintingCompositingClipPathPhase)) {
+        if (paintSVGClippingMask || (!(paintFlags & PaintLayerFlag::PaintingCompositingMaskPhase) && (paintFlags & PaintLayerFlag::PaintingCompositingClipPathPhase))) {
```
Indeed, a new PaintLayerFlag makes most sense.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4019,6 +4088,14 @@ RenderLayer* RenderLayer::enclosingFragmentedFlowAncestor() const
 
+RenderLayer* RenderLayer::enclosingSVGRootLayer() const
+{
+    RenderLayer* curr = parent();
```
s/RenderLayer/auto/?

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5414,6 +5491,30 @@ bool RenderLayer::isVisuallyNonEmpty(PaintedContentRequest* request) const
     return hasNonEmptyChildRenderers(*request);
 }
 
+RenderSVGResourceClipper* RenderLayer::svgClipperFromStyle() const
```
The code shouldn't live in RenderLayer, it doesn't need any RenderLayer related information. How about moving it to RenderLayerModelObject?

rwlbuis replied:
I am not sure we know where it belongs, it could belong in SVGResourcesCache and also there seems a small overlap with ReferencedSVGResources. I made it a FIXME but let me know if it should be somewhere else regardless.

nikolaszimmermann replied:
Good point, let's leave it as-is and decide later.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGImage.cpp:
```diff
@@ -132,8 +132,7 @@ void RenderSVGImage::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset)
     if (paintInfo.phase == PaintPhase::ClippingMask) {
-        // FIXME: [LBSE] Upstream SVGRenderSupport changes
-        // SVGRenderSupport::paintSVGClippingMask(*this, paintInfo);
+        SVGRenderSupport::paintSVGClippingMask(*this, paintInfo);
```
I know I put this into SVGRenderSupport in LBSE downstream, but I'd vote to move this into a shared class, maybe even RenderLayerModelObject?

rwlbuis replied:
Not sure, will it be used from non SVG? Added a FIXME for now.

nikolaszimmermann replied:
No, but RenderLayerModelOBject is the only common renderer that is the base of all SVG renderers.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -0,0 +1,309 @@
+void RenderSVGResourceContainer::idChanged()
+{
+    // Invalidate all our current clients.
+    removeAllClientsFromCache();
```
I'd say we should get rid of these calls, initially, no?

rwlbuis replied:
Right, removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -0,0 +1,309 @@
+    auto elements = copyToVectorOf<Ref<SVGElement>>(treeScope.removePendingSVGResource(m_id));
+
+    treeScope.addSVGResource(m_id, *this);
+
```
unnecessary blank

rwlbuis replied:
Removed.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -0,0 +1,309 @@
+
+RenderSVGResourceClipper::~RenderSVGResourceClipper() = default;
+
+SVGGraphicsElement* RenderSVGResourceClipper::shouldApplyPathClipping() const
```
This could live in the DOM, aka. SVGClipPathElement?

rwlbuis replied:
I moved it there, but maybe not so nice that SVGClipPathElement::shouldApplyPathClipping checks style as the first thing...

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -0,0 +1,309 @@
+    MaskClipping
+};
+
+static ClippingMode& currentClippingMode()
```
Hmm, the idea was to avoid switching from PathClipping when already performing MaskClipping...  We should find a nicer way then a shared static variable.

nikolaszimmermann replied:
Any idea on that one?

nikolaszimmermann replied:
Let's postpone a cleaner solution - this is at least the fastest, but it would be an issue if we ever do multi-threaded rendering.

rwlbuis replied:
Sure, I did not find a time to look into it but am not an expert on multi-threading.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
```diff
@@ -315,14 +315,12 @@ void RenderSVGRoot::paintObject(PaintInfo& paintInfo, const LayoutPoint& paintOf
 
     auto adjustedPaintOffset = paintOffset + location();
     if (paintInfo.phase == PaintPhase::Mask && style().visibility() == Visibility::Visible) {
-        // FIXME: [LBSE] Upstream SVGRenderSupport changes
```
You can leave it.

rwlbuis replied:
Yeah, I edited it too quickly, restored.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGPathData.h:
```diff
@@ -26,4 +26,6 @@ class Path;
 
 Path pathFromGraphicsElement(const SVGElement*);
 
+const Path& sharedClipAllPath();
```
What's the purpose of that?

rwlbuis replied:
The comment says "The SVG specification wants us to clip everything, if clip-path doesn't have a child."
I moved it into RenderSVGResourceClipper.cpp since that seems to be the only user.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -548,4 +549,19 @@ bool SVGHitTestCycleDetectionScope::isVisiting(const RenderElement& element)
+
+    ASSERT(renderer.isSVGLayerAwareRenderer());
+    const auto& referenceClipPathOperation = downcast<ReferencePathOperation>(*renderer.style().clipPath());
+    if (RenderSVGResourceContainer* renderResource = renderer.document().svgResourceById(referenceClipPathOperation.fragment())) {
```
auto?

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGResourcesCycleSolver.cpp:
```diff
@@ -46,7 +46,7 @@ bool SVGResourcesCycleSolver::resourceContainsCycles(LegacyRenderSVGResourceCont
     RenderObject* node = &resource;
     while (node) {
-        if (node != &resource && node->isSVGResourceContainer()) {
+        if (node != &resource && node->isLegacySVGResourceContainer()) {
```
Cycle solving is currently disabled for LBSE?

nikolaszimmermann replied:
Yes it is, and we're going to leave it disabled -- SVG resources are being reworked in parallel.

nikolaszimmermann reviewed (changes requested):
Wonderful, great to finally start upstreaming the resources :-)

nikolaszimmermann commented on Source/WebCore/dom/TreeScope.h:
```diff
@@ -135,7 +135,8 @@ class TreeScope {
     void addSVGResource(const AtomString& id, LegacyRenderSVGResourceContainer&);
     void removeSVGResource(const AtomString& id);
-    LegacyRenderSVGResourceContainer* svgResourceById(const AtomString& id) const;
+    RenderSVGResourceContainer* lookupSVGResourceById(const AtomString& id) const;
```
This needs `ENABLE(LAYER_BASED_SVG_ENGINE)` guards -- not all ports compile it by default yet IIRC.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/PaintPhase.h:
```diff
@@ -60,7 +60,7 @@ enum class PaintBehavior : uint32_t {
     ForceBlackText                      = 1 << 2,
     ForceWhiteText                      = 1 << 3,
-    RenderingSVGMask                    = 1 << 4,
+    RenderingSVGClipOrMask              = 1 << 4,
```
It's ok, but technically incorrect for legacy (where it is only used for masks).

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1534,6 +1535,16 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
+
+    auto hitsClipContent = [&](Element& element) -> bool {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+        if (auto* clipper = dynamicDowncast<RenderSVGResourceClipper>(element.renderer()))
```
The negation is either wrong or confusing - I guess the former.

rwlbuis replied:
I think it is wrong too, let's see the bots results.

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1534,6 +1535,16 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
+    auto hitsClipContent = [&](Element& element) -> bool {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+        if (auto* clipper = dynamicDowncast<RenderSVGResourceClipper>(element.renderer()))
+            return !clipper->hitTestClipContent(FloatRect(borderBoxRect()), hitTestLocationInLocalCoordinates);
```
Does it need implicit FloatRect conversion? Didn't check if that constructor taking a LayoutRect is 'explicit' or not. Either way, you could try { borderBoxRect() } instead in that case.

rwlbuis replied:
That seems to work.

nikolaszimmermann commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1549,8 +1560,7 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
             break;
         if (!is<SVGClipPathElement>(*element))
             break;
-        CheckedRef clipper = downcast<LegacyRenderSVGResourceClipper>(*element->renderer());
```
Oops, we lost the recent 'CheckedRef' introduction -- that needs to be fixed in your new hitsClipContent lambda!

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1592,6 +1593,16 @@ void RenderLayer::setAncestorChainHasVisibleDescendant()
 {
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+    if (renderer().document().settings().layerBasedSVGEngineEnabled()) {
+        m_enclosingSVGResourceContainer = lineageOfType<RenderSVGResourceContainer>(renderer()).first();
```
While this is convenient, we could halve the necessary tree walks, if we manually walk the ancestor chain once, and check for 'RenderSVGResourceContainer' and 'RenderSVGHiddenContainer' at the same time. I think this would be a good argument to do that instead of lineageOfType<>.

Also now that I think about it:

```
<defs>
  <clipPath>
    <rect>
```

The `<rect>` has both m_enclosingSVGResourceContainer and m_enclosingSVGHiddenContainer set. It would be sufficient to only have m_enclosingSVGResourceContainer set. In practice, with the current code, it won't make a difference, however for safety reasons this is a sane choice -- there is no intention to handle a specific case where both are set.

So I suggest to write the ancestor walk manually, and either stop at the first RenderSVGHiddenContainer or RenderSVGResourceContainer -- this way we avoid the two tree walks, and avoid awkward situations where both are set.

rwlbuis replied:
Done, the stop condition is a bit tricky.

nikolaszimmermann reviewed (commented):
Next round of comments, we're getting closer.

nikolaszimmermann commented on Source/WebCore/dom/TreeScope.h:
```diff
@@ -132,10 +132,15 @@ class TreeScope {
     void addSVGResource(const AtomString& id, RenderSVGResourceContainer&);
+#endif
     void addSVGResource(const AtomString& id, LegacyRenderSVGResourceContainer&);
     void removeSVGResource(const AtomString& id);
```
Spotted this: removeSVGResource hasn't been adapted -- it only removes from 'legacyResources' not 'resources'.

rwlbuis replied:
True! Fixed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3127,6 +3168,36 @@ void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSa
+            return;
+        }
+
         if (auto* clipperRenderer = ReferencedSVGResources::referencedClipperRenderer(renderer().treeScopeForSVGReferences(), referenceClipPathOperation)) {
```
That only supports legacy at present! referencedClipperRenderer returns a 'LegacyRenderSVGResourceClipper' - the actual LBSE implementation is still missing here.

nikolaszimmermann replied:
Not asking to implement this here in this PR, since it requires splitting up `ReferencedSVGResources` into a LBSE and non-LBSE variant, which deserves its own PR.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1592,6 +1593,26 @@ void RenderLayer::setAncestorChainHasVisibleDescendant()
+    m_enclosingSVGResourceContainer = nullptr;
+    m_enclosingSVGHiddenContainer = nullptr;
+    auto determineSVGAncestors = [&] (const RenderElement& renderer) {
+        for (auto* ancestor = renderer.parent(); ancestor; ancestor = ancestor->parent()) {
```
We started at 'renderer' before, not 'renderer.parent()' when we used lineageOfType. However, I do think this version is more correct.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1592,6 +1593,26 @@ void RenderLayer::setAncestorChainHasVisibleDescendant()
+        for (auto* ancestor = renderer.parent(); ancestor; ancestor = ancestor->parent()) {
+            if (auto* container = dynamicDowncast<RenderSVGResourceContainer>(ancestor)) {
+                m_enclosingSVGResourceContainer = container;
+                if (m_enclosingSVGHiddenContainer)
```
I'm confused by the exit condition. We wanted to avoid the situation where both are set.
I think it is sufficient to break when either m_enclosingSVGresourceContainer/HiddenContainer is non-null.

rwlbuis replied:
You are right, fixed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3096,7 +3138,7 @@ std::pair<Path, WindRule> RenderLayer::computeClipPath(const LayoutSize& offsetF
 }
 
-void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSaver& stateSaver, RegionContextStateSaver& regionContextStateSaver, const LayerPaintingInfo& paintingInfo, OptionSet<PaintLayerFlag> paintFlags, const LayoutSize& offsetFromRoot)
+void RenderLayer::setupClipPath(GraphicsContext& context, GraphicsContextStateSaver& stateSaver, RegionContextStateSaver& regionContextStateSaver, const LayerPaintingInfo& paintingInfo, OptionSet<PaintLayerFlag>& paintFlags, const LayoutSize& offsetFromRoot)
```
Nicer, than the bool reference.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5428,6 +5530,31 @@ bool RenderLayer::isVisuallyNonEmpty(PaintedContentRequest* request) const
+        return nullptr;
+
+    const auto& referenceClipPathOperation = downcast<ReferencePathOperation>(*renderer().style().clipPath());
+    if (RenderSVGResourceContainer* renderResource = renderer().document().lookupSVGResourceById(referenceClipPathOperation.fragment())) {
```
auto?

rwlbuis replied:
Sure, fixed.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -5428,6 +5530,31 @@ bool RenderLayer::isVisuallyNonEmpty(PaintedContentRequest* request) const
+
+    const auto& referenceClipPathOperation = downcast<ReferencePathOperation>(*renderer().style().clipPath());
+    if (RenderSVGResourceContainer* renderResource = renderer().document().lookupSVGResourceById(referenceClipPathOperation.fragment())) {
+        if (is<RenderSVGResourceClipper>(renderResource))
```
This is a good opportunity to use dynamicDowncast.

rwlbuis replied:
Done.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1322,6 +1337,12 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
     // Pointer to the enclosing RenderLayer that caused us to be paginated. It is 0 if we are not paginated.
     WeakPtr<RenderLayer> m_enclosingPaginationLayer;
 
+    // Pointer to the enclosing RenderSVGResourceContainer, if present.
```
guards missing.

rwlbuis replied:
Added.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
```diff
@@ -0,0 +1,233 @@
+    auto& context = paintInfo.context();
+    GraphicsContextStateSaver stateSaver(context);
+
+#if 0
```
Needs a // FIXME I guess.

rwlbuis replied:
Added.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGResourceClipper.h:
```diff
@@ -0,0 +1,67 @@
+
+    bool hitTestClipContent(const FloatRect&, const LayoutPoint&);
+
+    inline SVGUnitTypes::SVGUnitType clipPathUnits() const;
```
Dead code.

nikolaszimmermann replied:
Wrong, it's just not used consistently, there are places in RenderSVGResourceClipper that still use "clipPathElement().clipPathUnits()" instead of plain 'clipPathUnits()'.

rwlbuis replied:
I removed those instances.

nikolaszimmermann commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -657,4 +658,21 @@ FloatRect SVGRenderSupport::calculateApproximateStrokeBoundingBox(const RenderEl
+
+    ASSERT(renderer.isSVGLayerAwareRenderer());
+    const auto& referenceClipPathOperation = downcast<ReferencePathOperation>(*renderer.style().clipPath());
+    if (auto* renderResource = renderer.document().lookupSVGResourceById(referenceClipPathOperation.fragment())) {
```
Early exits could help readability here.

rwlbuis replied:
Done.

nikolaszimmermann reviewed (commented):
Great job @rwlbuis - not much left :-) Fortunately the first resource is the worst, the others are much less work....

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerInlines.h:
```diff
@@ -51,4 +52,19 @@ inline bool RenderLayer::paintsWithTransparency(OptionSet<PaintBehavior> paintBe
+    if (hasBlendMode() || isolatesBlending())
+        return true;
+
+    // SVG clip-paths may use clipping masks, if so, flag this layer as transparent.
```
Oops, this is also LBSE only, so this even needs the check condition if LBSE is enabled at run-time (via document().settings()...).

rwlbuis replied:
Right, fixed.

nikolaszimmermann reviewed (approved):
r=me, excellent job Rob! I've left another comment, still one thing to fix prior to mq+.

nikolaszimmermann committed:
[LBSE] Add RenderSVGResourceClipper
https://bugs.webkit.org/show_bug.cgi?id=262191

Reviewed by Nikolas Zimmermann.

Implement the first SVG resource in LBSE: clipping (RenderSVGResourceClipper).
Unlike the legacy engine, LBSE does not make use of temporary ImageBuffers for clipping,
but relies on Porter-Duff DestinationIn/SourceOver compositing operations, to achieve
the same effect, but faster. This aligns SVG clipping with HTML/CSS clipping.

This is only the beginning: dynamic invalidations are not working, as we're introducing
them incrementally in a unified, clean way.

Covered by existing tests (dozens of progressions, see TestExpectations).

* LayoutTests/platform/mac-ventura-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/animate-elem-32-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/coords-viewattr-03-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-05-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-elem-06-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/fonts-kern-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-intro-01-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-02-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-04-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/masking-path-05-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/painting-marker-02-f-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/shapes-intro-01-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-align-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-05-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/W3C-SVG-1.1/text-text-06-t-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textEffect2-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/batik/text/textProperties-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/carto.net/textbox-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/embedding-external-svgs-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/path-moveto-only-rendering-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/subpaths-moveto-only-rendering-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/text-clip-expected.png: Added.
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/text-clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/custom/use-on-symbol-inside-pattern-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/filters/filter-clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/foreignObject/clip-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/repaint/mask-clip-target-transform-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-align-01-b-expected.txt:
* LayoutTests/platform/mac-ventura-wk2-lbse-text/svg/text/text-text-05-t-expected.txt:
* LayoutTests/svg/clip-path/clip-opacity.html:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/TreeScope.cpp:
(WebCore::TreeScope::removeSVGResource):
(WebCore::TreeScope::lookupSVGResourceById const):
(WebCore::TreeScope::lookupLegacySVGResoureById const):
(WebCore::TreeScope::svgResourceById const): Deleted.
* Source/WebCore/dom/TreeScope.h:
* Source/WebCore/rendering/PaintPhase.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::hitTestClipPath const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
(WebCore::RenderLayer::updateAncestorDependentState):
(WebCore::RenderLayer::computeHasVisibleContent const):
(WebCore::RenderLayer::beginTransparencyLayers):
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::setupClipPath):
(WebCore::RenderLayer::paintLayerContents):
(WebCore::RenderLayer::enclosingSVGRootLayer const):
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::isPaintingSVGResourceLayer const):
* Source/WebCore/rendering/RenderLayerInlines.h:
(WebCore::RenderLayer::hasFilter const):
(WebCore::RenderLayer::hasNonOpacityTransparency const):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isSVGForeignObject const):
(WebCore::RenderObject::isSVGResourceContainer const):
(WebCore::RenderObject::isLegacySVGResourceClipper const):
(WebCore::RenderObject::isSVGResourceClipper const):
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::paint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::paint):
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::computeClipPath const):
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGResource.cpp:
(WebCore::requestPaintingResource):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp: Added.
(WebCore::RenderSVGResourceClipper::RenderSVGResourceClipper):
(WebCore::RenderSVGResourceClipper::shouldApplyPathClipping const):
(WebCore::currentClippingMode):
(WebCore::sharedClipAllPath):
(WebCore::RenderSVGResourceClipper::applyPathClipping):
(WebCore::RenderSVGResourceClipper::applyMaskClipping):
(WebCore::RenderSVGResourceClipper::hitTestClipContent):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.h: Added.
* Source/WebCore/rendering/svg/RenderSVGResourceClipperInlines.h: Added.
(WebCore::RenderSVGResourceClipper::clipPathElement const):
(WebCore::RenderSVGResourceClipper::clipPathUnits const):
* Source/WebCore/rendering/svg/RenderSVGResourceSolidColor.cpp:
(WebCore::RenderSVGResourceSolidColor::applyResource):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::paintObject):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::paint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::paint):
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::SVGInlineTextBox::paint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::paintSVGClippingMask):
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::isRenderingMaskImage):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::drawContentIntoMaskImage):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.h:
(isType):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.h:
(WebCore::getRenderSVGResourceContainerById):
* Source/WebCore/svg/SVGClipPathElement.cpp:
(WebCore::SVGClipPathElement::createElementRenderer):
(WebCore::SVGClipPathElement::shouldApplyPathClipping const):
* Source/WebCore/svg/SVGClipPathElement.h:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::isSVGLayerAwareElement):

Canonical link: https://commits.webkit.org/269635@main
